appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
           node.public_ip = self.login_host
 
     if self.disks:
-      valid, reason = self.is_disks_valid(nodes)
-      if not valid:
-        return self.invalid(reason)
+      self.validate_disks(nodes)
 
       for node in nodes:
         node.disk = self.disks.get(node.public_ip)
 
     rep = self.is_database_replication_valid(nodes)
 
-    if not rep['result']:
+    if not rep:
       return rep
 
     self.nodes = nodes
-    return self.valid()
-
+    return True
 
-  def is_valid_advanced_format(self):
-    """Checks to see if this NodeLayout represents an acceptable advanced
+  # TODO: Deprecated format, remove when we switch to one ips_layout format.
+  def is_valid_node_format(self):
+    """Checks to see if this NodeLayout represents an acceptable (new) advanced
     deployment strategy, and if so, constructs self.nodes from it.
-
+    
     Returns:
-      A dict that indicates if the deployment strategy is valid, and if
-      not, the reason why it is invalid.
+      True if the deployment strategy is valid.
+    Raises:
+      BadConfigurationException with reason if the deployment strategy is not 
+        valid.
     """
-    if self.nodes:
-      return self.valid()
-
+    # Keep track of whether the deployment is valid while going through.
     node_hash = {}
+    role_count = {
+      'appengine': 0,
+      'master': 0,
+      'memcache': 0,
+      'taskqueue': 0,
+      'zookeeper': 0,
+      'login': 0
+    }
+    node_count = 1
+    db_master_created = False
+    tq_master_created = False
+    master_node = None
+    login_found = False
+    # Loop through the list of "node sets", which are grouped by role.
     for role, ips in self.input_yaml.iteritems():
-      
-      if isinstance(ips, str):
-        ips = [ips]
+      if len(ips) == 0:
+        self.invalid("Node amount cannot be zero.")
+      ips_list = ips if isinstance(ips, list) else [ips]
+      test_ips = [self.is_cloud_ip(ip) for ip in ips_list]
+      cloud = all(test_ips) or not any(test_ips)
```

Comment:
test_ips is a list of boolean values because is_cloud_ip returns a boolean value, so they should all be true if we were given node-id and none should be true if we were given ip addresses. This does mean we can't do what we're currently able to do where we can use node-id and ip addresses at the same time. If we want to keep supporting the ability to use them both together it could just be `cloud = any(test_ips)`